src/brod_gssapi.erl

```diff
@@ -5,134 +5,110 @@
-                                send_sasl_token(Token, Sock, Mod),
-                                SaslRes;
-                            true ->
-                                SaslRes
-                        end
-                                 end,
-                    case do_while(CliStepFun, CondFun) of
-                        ?SASL_OK ->
-                            ok = setopts(Sock, Mod, [{active, once}]);
-                        ?SASL_CONTINUE ->
-                            sasl_recv(Mod, Sock, Timeout);
-                        Other ->
-                            {error, Other}
-                    end
-            end;
-        {error, closed} ->
-            {error, bad_credentials};
-        Unexpected ->
-            {error, Unexpected}
-    end.
+% Observed Sequence of handshake is as follows:
+-spec auth(
+    Host :: string(),
+    Sock :: gen_tcp:socket() | ssl:sslsocket(),
+    HandshakeVsn :: non_neg_integer(),
+    Mod :: gen_tcp | ssl,
+    ClientId :: binary(),
+    Timeout :: pos_integer(),
+    SaslOpts :: term()
+) -> ok | {error, Reason :: term()}.
+auth(
+    Host,
+    Sock,
+    HandshakeVsn,
+    Mod,
+    ClientId,
+    Timeout,
+    Opts
+) ->
+    State = new(Host, Sock, HandshakeVsn, Mod, ClientId, Timeout, Opts),
+    dispatch(State).
 
-%%====================================================================
-%% Internal functions
-%%====================================================================
+dispatch(#{handshake_vsn := 1} = State) ->
+    brod_gssapi_v1:auth(State);
+dispatch(#{handshake_vsn := 0} = State) ->
+    brod_gssapi_v0:auth(State);
+dispatch(_State) ->
+    {error, undefined_handshake_vsn}.
 
--spec ensure_binary(atom() | iodata()) -> binary().
-ensure_binary(Atom) when is_atom(Atom) -> atom_to_binary(Atom, utf8);
-ensure_binary(Str) -> iolist_to_binary(Str).
-
-do_while(Fun, CondFun) ->
-    case CondFun(Fun()) of
-        continue -> do_while(Fun, CondFun);
-        Other -> Other
-    end.
+%% Backwards compat where handshake isn't given
+%auth(Host, Sock, Mod, _ClientId, Timeout, _SaslOpts = {_Method = gssapi, Keytab, Principal}) ->
+%
 
-setopts(Sock, _Mod = gen_tcp, Opts) -> inet:setopts(Sock, Opts);
-setopts(Sock, _Mod = ssl, Opts)     ->  ssl:setopts(Sock, Opts).
+-spec new(
+    Host :: string(),
+    Sock :: gen_tcp:socket() | ssl:sslsocket(),
+    HandshakeVsn :: non_neg_integer(),
+    Mod :: gen_tcp | ssl,
+    ClientId :: binary(),
+    Timeout :: pos_integer(),
+    SaslOpts :: term()
+) -> state().
+new(Host, Sock, HandshakeVsn, Mod, ClientId, Timeout, {Method, KeyTab, Principal}) ->
+    #{
+        host => ensure_binary(Host),
+        sock => Sock,
+        transport_mod => Mod,
+        client_id => ClientId,
+        timeout => Timeout,
+        method => Method,
+        mechanism => <<"GSSAPI">>,
+        keytab => ensure_binary(KeyTab),
+        principal => ensure_binary(Principal),
+        sasl_context => <<"kafka">>,
```

Comment:
Can hard-coding this cause some problem?

Reply:
It should not be hard coded. We should use macros and/or functions.